src/Datadog.Trace/Configuration/TracerSettings.cs

```diff
@@ -286,31 +286,50 @@ public static CompositeConfigurationSource CreateDefaultConfigurationSource()
             return GlobalSettings.CreateDefaultConfigurationSource();
         }
 
-        internal bool IsIntegrationEnabled(string name)
+        /// <summary>
+        /// Populate the internal structures. Modifying the settings past this point is not supported
+        /// </summary>
+        internal void Freeze()
+        {
+            Integrations.SetDisabledIntegrations(DisabledIntegrationNames);
+        }
+
+        internal bool IsIntegrationEnabled(int integrationId)
         {
             if (TraceEnabled && !DomainMetadata.ShouldAvoidAppDomain())
             {
-                bool disabled = Integrations[name].Enabled == false || DisabledIntegrationNames.Contains(name);
-                return !disabled;
+                bool? enabled = Integrations[integrationId].Enabled;
+                return enabled != false;
             }
 
             return false;
         }
 
-        internal bool IsOptInIntegrationEnabled(string name)
+        internal bool IsIntegrationEnabled(string integrationName)
         {
             if (TraceEnabled && !DomainMetadata.ShouldAvoidAppDomain())
             {
-                var disabled = Integrations[name].Enabled != true || DisabledIntegrationNames.Contains(name);
-                return !disabled;
+                bool? enabled = Integrations[integrationName].Enabled;
+                return enabled != false;
             }
 
             return false;
         }
 
-        internal double? GetIntegrationAnalyticsSampleRate(string name, bool enabledWithGlobalSetting)
+        internal bool IsOptInIntegrationEnabled(int integrationId)
```

Comment:
For readability, is it worth refactoring this opt-in vs opt-out business by removing this method and adding a new parameter to the `IsIntegrationEnabled()` overloads?

```csharp
IsIntegrationEnabled(int integrationId, bool defaultValue = true)
{
    if (TraceEnabled && !DomainMetadata.ShouldAvoidAppDomain())
    {
        // if not explicitly set, fall back to `defaultValue`
        return Integrations[integrationId].Enabled ?? defaultValue;
    }
}
```

...and replacing the one call to `IsOptInIntegrationEnabled(id)` with `IsOptInIntegrationEnabled(id, false)`?

Reply:
Actually, we should probably hold off on this since it causes a breaking change if integrations load an older `Datadog.Trace.dll`.

Reply:
Done